presto-main/src/main/java/io/prestosql/sql/planner/optimizations/UnaliasSymbolReferences.java

```diff
@@ -126,695 +122,988 @@ public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, Sym
+            PlanAndMappings rewrittenSource = node.getSource().accept(this, context);
+            SymbolMapper mapper = new SymbolMapper(rewrittenSource.getSymbolMappings());
 
-            List<Symbol> canonical = Lists.transform(node.getOutputSymbols(), this::canonicalize);
-            return new OutputNode(node.getId(), source, node.getColumnNames(), canonical);
+            OrderingScheme newOrderingScheme = mapper.map(node.getOrderingScheme());
+
+            return new PlanAndMappings(
+                    new SortNode(node.getId(), rewrittenSource.getRoot(), newOrderingScheme, node.isPartial()),
+                    mapper.getMapping());
         }
 
         @Override
-        public PlanNode visitEnforceSingleRow(EnforceSingleRowNode node, RewriteContext<Void> context)
+        public PlanAndMappings visitFilter(FilterNode node, UnaliasContext context)
         {
-            PlanNode source = context.rewrite(node.getSource());
+            PlanAndMappings rewrittenSource = node.getSource().accept(this, context);
+            SymbolMapper mapper = new SymbolMapper(rewrittenSource.getSymbolMappings());
+
+            Expression newPredicate = mapper.map(node.getPredicate());
 
-            return new EnforceSingleRowNode(node.getId(), source);
+            return new PlanAndMappings(
+                    new FilterNode(node.getId(), rewrittenSource.getRoot(), newPredicate),
+                    mapper.getMapping());
         }
 
         @Override
-        public PlanNode visitAssignUniqueId(AssignUniqueId node, RewriteContext<Void> context)
+        public PlanAndMappings visitProject(ProjectNode node, UnaliasContext context)
         {
-            PlanNode source = context.rewrite(node.getSource());
+            PlanAndMappings rewrittenSource = node.getSource().accept(this, context);
+            SymbolMapper mapper = new SymbolMapper(rewrittenSource.getSymbolMappings());
+
+            // canonicalize ProjectNode assignments
+            ImmutableList.Builder<Map.Entry<Symbol, Expression>> builder = ImmutableList.builder();
+            for (Map.Entry<Symbol, Expression> assignment : node.getAssignments().entrySet()) {
+                builder.add(new SimpleEntry<>(mapper.map(assignment.getKey()), mapper.map(assignment.getValue())));
+            }
+            List<Map.Entry<Symbol, Expression>> rewrittenAssignments = builder.build();
 
-            return new AssignUniqueId(node.getId(), source, node.getIdColumn());
+            // deduplicate assignments
+            Map<Symbol, Expression> deduplicateAssignments = new HashMap<>();
+            for (Map.Entry<Symbol, Expression> assignment : rewrittenAssignments) {
+                Expression previous = deduplicateAssignments.put(assignment.getKey(), assignment.getValue());
+                checkState(previous == null || assignment.getValue().equals(previous), "different expressions projected to the same symbol");
+            }
```

Comment:
Cannot use `mapper` in lambda, because it is not effectively final. Instead, I used the Collector suggested by you for deduplication.

Reply:
You could make it effectively final by not reusing the same variable further down this method.

Reply:
The variable is reused as a leftover of the previous approach where mapper was modifiable. I intentionally kept it this way so that we can easily go back to that idea.